Modify reward per second

sujine2 opened the pull request:
- rewardPerSecond is changed.
- owner can set rewardPerSecond

1. When opening the pool for the first time, set the rewardPerSeed and duration.
2. The owner puts the next reward in before the duration.
3. After duration,  the rewardPerSecond is set  automatically based on the reward that owner put.
4. Repeat the above 2, 3

Considerations
- Owner can set nextRewardAmount variable.
- Owner cant set duration.

sujine2 committed:
modify rewardPerSecond

sujine2 committed:
change rewardPerSecond

sujine2 committed:
add change rewardPerSecond stake test code

sujine2 committed:
add change rewardPerSecond claim test code

sujine2 committed:
add change rewardPerSecond withdraw test code

sujine2 committed:
delete unused imports

sujine2 committed:
remove migrate

sujine2 commented on contracts/StakingPoolV2.sol:
```diff
@@ -131,7 +134,47 @@ contract StakingPoolV2 is IStakingPoolV2, StakedElyfiToken {
   }
 
-  // TODO Implement `migrate` function to send an asset to the next staking contract
+  /// @notice Migrate the amount of principal to the current round and transfer the rest principal to the caller
```
Please ignore this function

sujine2 commented on contracts/logic/StakingPoolLogicV2.sol:
```diff
@@ -39,24 +39,34 @@ library StakingPoolLogicV2 {
     return result;
   }
 
+  function updateRewardPerSecond(StakingPoolV2.PoolData storage poolData, address user) internal {
```
This is the most important part of the rewardPerSecond increase. 
Please review it.

sujine2 replied:
This is the most important function of the rewardPerSecond increase

sujine2 commented on contracts/StakingPoolV2.sol:
```diff
@@ -184,15 +227,14 @@ contract StakingPoolV2 is IStakingPoolV2, StakedElyfiToken {
     uint256 rewardPerSecond,
     uint256 startTimestamp,
-    uint256 duration
+    uint32 duration
```
Please check changing data type.

qpzm replied:
There is no need to replace uint256 into uint32 because both use 32bytes. Storage variables are packed, but this is not true in memory and calldata.

> Among others, the ABI specification requires arguments to be padded to multiples of 32 bytes. 

Reference: https://docs.soliditylang.org/en/v0.8.4/internals/layout_in_calldata.html

qpzm replied:
```suggestion
    uint256 duration
```

sujine2 commented on contracts/interface/IStakingPoolV2.sol:
```diff
@@ -33,12 +33,16 @@ interface IStakingPoolV2 {
     uint256 endTimestamp
   );
 
+  event Migrate(address user, uint256 amount);
```
Please ignore this event

sujine2 commented on contracts/interface/IStakingPoolV2.sol:
```diff
@@ -33,12 +33,16 @@ interface IStakingPoolV2 {
 
   function withdraw(uint256 amount) external;
 
+  function migrate(uint256 amount) external;
```
Please ignore this function

sujine2 commented on test/stake.test.ts:
```diff
@@ -235,4 +245,65 @@ describe('StakingPool.stake', () => {
   });
+
+
+  context('change rewardPerSecond', async () => {
```
Main test function in this commit

sujine2 commented on test/stake.test.ts:
```diff
@@ -261,22 +262,29 @@ describe('StakingPool.stake', () => {
     });
     
-    it.only('rewardPerSecond and stake in pool', async () => {
+    it('rewardPerSecond is changed and stake in pool', async () => {
```
Main test function in stake test

sujine2 commented on test/utils/expect.ts:
```diff
@@ -30,6 +30,40 @@ export function expectDataAfterStake(
   return [newPoolData, newUserData];
 }
 
+export function expectDataAfterStakeFixedRewardPersecond(
```
Main test function in stake test (change function name expectDataAfterStakeFixedRewardPersecond -> expectDataAfterStakeSetReward)

sujine2 commented on test/claim.test.ts:
```diff
@@ -196,30 +178,114 @@ describe('StakingPool.claim', () => {
-      
     // });
+
+    context('change rewardPerSecond', async () => {
```
Main test function in claim test

sujine2 commented on test/utils/expect.ts:
```diff
@@ -30,7 +30,7 @@ export function expectDataAfterStake(
 }
 
-export function expectDataAfterStakeFixedRewardPersecond(
+export function expectDataAfterStakeSetReward(
```
Main test function in stake test

sujine2 commented on test/utils/expect.ts:
```diff
@@ -119,6 +120,34 @@ export function expectDataAfterClaim(
   return [newPoolData, newUserData];
 }
 
+export function expectDataAfterClaimSetReward(
```
Main test function in claim test

sujine2 commented on contracts/StakingPoolV2.sol:
```diff
@@ -204,8 +204,10 @@ contract StakingPoolV2 is IStakingPoolV2, StakedElyfiToken {
   }
 
   function _claim(address user) internal {
```
Please check the full index update

sujine2 commented on contracts/StakingPoolV2.sol:
```diff
@@ -204,8 +204,10 @@ contract StakingPoolV2 is IStakingPoolV2, StakedElyfiToken {
   }
 
   function _claim(address user) internal {
+    if(_poolData.isOpened == true && block.timestamp > _poolData.endTimestamp) {
```
Please check the full index update

sujine2 commented on contracts/logic/StakingPoolLogicV2.sol:
```diff
@@ -63,7 +63,7 @@ library StakingPoolLogicV2 {
     address user
   ) internal {
-    if(block.timestamp > poolData.endTimestamp) {
+    if(poolData.isOpened == true && block.timestamp > poolData.endTimestamp) {
```
Please check the conditions under which calls the function.

donguks replied:
Is there no problem without updating another user index? I think, users who don't update the index can claim more or less reward.

donguks replied:
I think, this implementations can return invalid getRewardIndex. If there is no transactions like stake, `endTimestamp` is not up to date and `getRewardIndex()` return previous pool's reward index.

sujine2 commented on test/utils/expect.ts:
```diff
@@ -96,11 +95,48 @@ export function expectDataAfterWithdraw(
 }
 
-export function expectDataAfterClaim(
+export function expectDataAfterWithdrawSetReward(
```
Main test function in withdraw test

sujine2 commented on test/withdraw.test.ts:
```diff
@@ -265,5 +253,132 @@ describe('StakingPool.withdraw', () => {
     });
+
+
+    context('change rewardPerSecond', async () => {
```
Main test function in withdraw test

jaeboklee-modo commented on test/settings.test.ts:
```diff
@@ -138,4 +92,4 @@ describe('StakingPool.settings', () => {
     });
   });
-});
+});
```
Can you add the test case about the added functions? like inputNextReward or closePool.

jaeboklee-modo commented on test/claim.test.ts:
```diff
@@ -114,37 +111,179 @@ describe('StakingPool.claim', () => {
+        
+        const claimTx = await testEnv.stakingPool.connect(alice).claim();
+        const [expectedPoolData_2, expectedUserData_2] = expectDataAfterClaimSetReward(
+          expectedPoolData_1,
```
Why does this function get `expectedPoolData_1` and `expectedUserData_1`? I think it should take actual pool data from the current contract state as parameters.

jaeboklee-modo commented on test/claim.test.ts:
```diff
@@ -114,37 +111,179 @@ describe('StakingPool.claim', () => {
+
+        const stakeTx = await testEnv.stakingPool.connect(alice).stake(stakeAmount);
+        const [expectedPoolData_2, expectedUserData_2] = expectDataAfterStakeSetReward(
+          expectedPoolData_1,
```
It is the same as above.

jaeboklee-modo commented on test/stake.test.ts:
```diff
@@ -201,4 +244,95 @@ describe('StakingPool.stake', () => {
+      );
+      
+      const stakeTx_1 = await testEnv.stakingPool.connect(alice).stake(stakeAmount);
+      const [expectedPoolData_2, expectedUserData_2] = expectDataAfterStakeSetReward(
```
https://github.com/elysia-dev/elyfi-stakingpool-v2/pull/6/files#diff-a497d96c45659b1c0ac18979682159d45ed5ae9cf3803fe3161078fb2bc40ca0R234

Same thing!

jaeboklee-modo commented on test/stake.test.ts:
```diff
@@ -201,4 +244,95 @@ describe('StakingPool.stake', () => {
+      const stakeTx_2 = await testEnv.stakingPool.connect(alice).stake(stakeAmount);
+
+      const [expectedPoolData_3, expectedUserData_3] = expectDataAfterStake(
+        expectedPoolData_2,
```
After these `expectedPoolData_` parameter issues are solved, I'd review the rest code and logic.

qpzm commented on contracts/logic/StakingPoolLogicV2.sol:
```diff
@@ -42,40 +41,54 @@ library StakingPoolLogicV2 {
   }
 
+  function updateRewardPerSecond(StakingPoolV2.PoolData storage poolData, address user) internal {
+    poolData.userReward[user] = getUserReward(poolData, user);
+    poolData.rewardIndex = poolData.userIndex[user] = getRewardIndex(poolData);
```
There is no need to update `userReward` and `userIndex` because it does not change any user's status but affects only the pool status.

qpzm replied:
This also passes the all tests.
```suggestion
    poolData.rewardIndex = getRewardIndex(poolData);
```

sujine2 committed:
merge expectData test function

sujine2 committed:
Use eql to check keys with different values

Reference
https://www.chaijs.com/api/bdd/#method_eql

sujine2 committed:
fix test code error

sujine2 committed:
improve updateRewardPerSecond function

Co-authored-by: Hyunmin Lee <lhm1442@gmail.com>

sujine2 committed:
improve updateRewardPerSecond function and check passing test code

sujine2 committed:
resolve conflict

qpzm committed:
Delete unused PoolData.rewardPerSecondReduceRate

qpzm committed:
Fix lint

qpzm committed:
Change the input parameter type: uint32 -> uint256

There is no need to replace uint256 into uint32 because both use 32bytes. Storage variables are packed, but this is not true in memory and calldata.

Co-authored-by: sujine2 <52918738+sujine2@users.noreply.github.com>

qpzm committed:
Fix lint

qpzm commented on contracts/logic/StakingPoolLogicV2.sol:
```diff
@@ -45,16 +45,29 @@ library StakingPoolLogicV2 {
+
+    poolData.startTimestamp = poolData.lastUpdateTimestamp = poolData.endTimestamp;
+    poolData.endTimestamp = poolData.endTimestamp + poolData.duration;
+    poolData.rewardPerSecond = poolData.nextRewardAmount / poolData.duration;
```
[Note] `rewardPerSecond` is dependent on `poolData.duration`.

qpzm commented on test/utils/helpers.ts:
```diff
@@ -36,3 +36,39 @@ export const getPoolData = async (testEnv: TestEnv) => {
 };
+
+
+export const getSumUserData = async (
```
This function is not used anywhere. What is the purpose of this?

donguks commented on contracts/StakingPoolV2.sol:
```diff
@@ -205,6 +206,11 @@ contract StakingPoolV2 is IStakingPoolV2, StakedElyfiToken {
     _poolData.isFinished = true;
   }
 
+  function inputNextReward(uint256 amount) external onlyAdmin {
```
I think, `setNextReward` is better.

donguks commented on test/claim.test.ts:
```diff
@@ -15,50 +15,31 @@ require('./utils/matchers.ts');
-    thirdRoundInit.day,
-    BigNumber.from(10)
-  );
+  const rewardPersecond = BigNumber.from(utils.parseEther('1'));
```
We can make below variable declarations simple by using [moment](https://momentjs.com/).
Also we do not need type-casting to BigNumber because initNewPool's timestamp is [BigNumberish](https://docs.ethers.io/v5/api/utils/bignumber/#BigNumberish). 
```
const firstRoundTimestamp = moment('2022.7.7 10:00:00', 'YYYY.MM.DD hh:mm:ss').unix()
```

donguks commented on contracts/StakingPoolV2.sol:
```diff
@@ -205,6 +206,11 @@ contract StakingPoolV2 is IStakingPoolV2, StakedElyfiToken {
+    SafeERC20.safeTransferFrom(rewardAsset, msg.sender, address(this), amount);
+  }
+
   function retrieveResidue() external onlyAdmin {
```
Because rewardAsset and stakingAsset is same, `retrieveResidue` can withdraw all staked. It can be evil.
 I think, the method should retrieve only rewardAsset.

donguks commented on contracts/StakingPoolV2.sol:
```diff
@@ -22,12 +22,14 @@ contract StakingPoolV2 is IStakingPoolV2, StakedElyfiToken {
   }
```
Is there any ownership transfer method?

donguks commented on contracts/StakingPoolV2.sol:
```diff
@@ -160,8 +160,10 @@ contract StakingPoolV2 is IStakingPoolV2, StakedElyfiToken {
   }
 
   function _claim(address user) internal {
```
I think, we'd better add an emergency stop. When emergency stop is true, `_claim` is not work and we can call `retrieveResidue`. With this feature, we can resolve the wrong rewardPerSeconds issue which occurs occasionally.

sujine2 committed:
change function name and eidt witdraw amount in retrieveResidue function

sujine2 committed:
change function name and eidt witdraw amount in retrieveResidue function

qpzm committed:
Use evm_mine `timestamp` parameter

Reference
https://github.com/trufflesuite/ganache/pull/13/files

qpzm committed:
Add await when testing events

qpzm committed:
Add a test for retrieveResidue

jaeboklee-modo commented on test/utils/expect.ts:
```diff
@@ -1,47 +1,23 @@
   userData: UserData,
   txTimeStamp: BigNumber,
   amount: BigNumber,
   rewardPerSecond: BigNumber,
```
I'm curious whether rewardPerSecond and duration in function parameter are different with these two in poolData.
